controller_manager/src/controller_manager.cpp

```diff
@@ -53,61 +64,772 @@ ControllerManager::load_controller(
+  std::lock_guard<std::mutex> guard(services_lock_);
+  RCLCPP_DEBUG(get_logger(), "loading service locked");
+
+  response->ok = load_controller(request->name).get();
+
+  RCLCPP_DEBUG(
+    get_logger(), "loading service finished for controller '%s' ",
+    request->name.c_str());
+}
+
+void ControllerManager::reload_controller_libraries_service_cb(
+  const std::shared_ptr<controller_manager_msgs::srv::ReloadControllerLibraries::Request> request,
+  std::shared_ptr<controller_manager_msgs::srv::ReloadControllerLibraries::Response> response)
+{
+  // lock services
+  RCLCPP_DEBUG(get_logger(), "reload libraries service called");
+  std::lock_guard<std::mutex> guard(services_lock_);
+  RCLCPP_DEBUG(get_logger(), "reload libraries service locked");
+
+  // only reload libraries if no controllers are running
+  std::vector<std::string> loaded_controllers, running_controllers;
+  loaded_controllers = get_controller_names();
+  {
+    // lock controllers
+    std::lock_guard<std::recursive_mutex> guard(rt_controllers_wrapper_.controllers_lock_);
+    for (const auto & controller : rt_controllers_wrapper_.get_updated_list(guard)) {
+      if (is_controller_running(*controller.c)) {
+        running_controllers.push_back(controller.info.name);
+      }
+    }
+  }
+  if (!running_controllers.empty() && !request->force_kill) {
+    RCLCPP_ERROR(
+      get_logger(), "Controller manager: Cannot reload controller libraries because"
+      " there are still %i controllers running",
+      (int)running_controllers.size());
+    response->ok = false;
+    return;
+  }
+
+  // stop running controllers if requested
+  if (!loaded_controllers.empty()) {
+    RCLCPP_INFO(get_logger(), "Controller manager: Stopping all running controllers");
+    std::vector<std::string> empty;
+    if (switch_controller(
+        empty, running_controllers,
+        controller_manager_msgs::srv::SwitchController::Request::BEST_EFFORT) !=
+      controller_interface::return_type::SUCCESS)
+    {
+      RCLCPP_ERROR(
+        get_logger(),
+        "Controller manager: Cannot reload controller libraries because failed to stop "
+        "running controllers");
+      response->ok = false;
+      return;
+    }
+    for (const auto & controller : loaded_controllers) {
+      if (unload_controller(controller) != controller_interface::return_type::SUCCESS) {
+        RCLCPP_ERROR(
+          get_logger(), "Controller manager: Cannot reload controller libraries because "
+          "failed to unload controller '%s'",
+          controller.c_str());
+        response->ok = false;
+        return;
+      }
+    }
+    loaded_controllers = get_controller_names();
+  }
+  assert(loaded_controllers.empty());
 
-  loaded_controllers_.emplace_back(controller);
-  return loaded_controllers_.back();
+  // Force a reload on all the PluginLoaders (internally, this recreates the plugin loaders)
+  for (const auto & controller_loader : loaders_) {
+    controller_loader->reload();
```

Comment:
shouldn't the controller be `re-configured` as well?

Reply:
In the old version it wasn't: https://github.com/ros-controls/ros_control/blob/noetic-devel/controller_manager/src/controller_manager.cpp#L663

We could discuss adding it, after reloading there are no loaded controllers. We may want to load (create + configure) them.

Reply:
I'd err on the side of caution here, we don't know what we are loading and whether the newly loaded ones will be compatible with each other.